### PR TITLE
Drag and drop forms group collapsed

### PR DESCRIPTION
**Pull request by elpaso**
## Description

This PR adds a new configuration options for D&D form group boxes: it is now possible to collapse the groups and control the collapsed status with an expression:

image

Funded by: ARPA Piemonte

### CODE REVIEW

File: python/core/auto_generated/editform/qgsattributeeditorcontainer.sip.in [hunk 1]
```diff
@@ -112,6 +134,30 @@ show or hide this container based on an expression incorporating
 the field value controlled by editor widgets.
 
 .. versionadded:: 3.0
+%End
+
+    QgsOptionalExpression collapsedExpression() const;
```
Comment (nyalldawson) [inline python/core/auto_generated/editform/qgsattributeeditorcontainer.sip.in:139]: This should really use QgsPropertyCollection, but I can see that you've modelled it off the existing visibility expression, so you get a pass here...

(maybe we should sip_skip it though to make a future move to QgsPropertyCollection easier)

File: src/core/editform/qgsattributeeditorcontainer.h [hunk 2]
```diff
@@ -66,6 +66,25 @@ class CORE_EXPORT QgsAttributeEditorContainer : public QgsAttributeEditorElement
      */
     virtual bool isGroupBox() const { return mIsGroupBox; }
 
+    /**
+     * For containers rendedered a group box returns if this group box is collapsed.
```
Comment (nyalldawson) [inline src/core/editform/qgsattributeeditorcontainer.h:70]:
```suggestion
     * For group box containers returns if this group box is collapsed.
```

File: src/core/editform/qgsattributeeditorcontainer.h [hunk 3]
```diff
@@ -66,6 +66,25 @@ class CORE_EXPORT QgsAttributeEditorContainer : public QgsAttributeEditorElement
...
+     *
+     * \returns TRUE if the group box, FALSE otherwise.
```
Comment (nyalldawson) [inline src/core/editform/qgsattributeeditorcontainer.h:72]:
```suggestion
     * \returns TRUE if the group box is collapsed, FALSE otherwise.
```

File: src/core/editform/qgsattributeeditorcontainer.h [hunk 4]
```diff
@@ -66,6 +66,25 @@ class CORE_EXPORT QgsAttributeEditorContainer : public QgsAttributeEditorElement
...
+     * \see collapsed()
+     * \see setCollapsed()
+     * \since QGIS 3.26
+     */
+    bool collapsed() const { return mCollapsed; };
+
+    /**
+     * For containers rendedered a group box sets if this group box is \a collapsed.
```
Comment (nyalldawson) [inline src/core/editform/qgsattributeeditorcontainer.h:80]:
```suggestion
     * For group box containers sets if this group box is \a collapsed.
```

File: src/core/editform/qgsattributeeditorcontainer.h [hunk 5]
```diff
@@ -127,6 +146,28 @@ class CORE_EXPORT QgsAttributeEditorContainer : public QgsAttributeEditorElement
      */
     void setVisibilityExpression( const QgsOptionalExpression &visibilityExpression );
 
+    /**
+     * The collapsed expression is used in the attribute form to
+     * set the collapsed status of the group box container container based on an expression incorporating
+     * the field value controlled by editor widgets. This property is ignored if the container is not
+     * rendered as a group box.
```
Comment (nyalldawson) [inline src/core/editform/qgsattributeeditorcontainer.h:153]:
```suggestion
     * a group box.
```

File: src/core/editform/qgsattributeeditorcontainer.h [hunk 6]
```diff
@@ -127,6 +146,28 @@ class CORE_EXPORT QgsAttributeEditorContainer : public QgsAttributeEditorElement
...
+     * \see setCollapsedExpression()
+     * \since QGIS 3.26
+     */
+    QgsOptionalExpression collapsedExpression() const;
+
+    /**
+     * The collapsed expression is used in the attribute form to
+     * set the collapsed status of the group box of this container based on an expression incorporating
+     * the field value controlled by editor widgets. This property is ignored if the container is not
+     * rendered as a group box.
```
Comment (nyalldawson) [inline src/core/editform/qgsattributeeditorcontainer.h:164]:
```suggestion
     * a group box.
```

(rendered sounds odd here!)

File: tests/src/python/test_qgsrelation.py [hunk 7]
```diff
@@ -196,6 +196,13 @@ def testValidRelationAfterChangingStyle(self):
         self.assertTrue(valid)
 
         # update style
+        # Note: the project is re-read because of a subtle bug with bidings involving
```
Comment (nyalldawson) [inline tests/src/python/test_qgsrelation.py:199]:
```suggestion
        # Note: the project is re-read because of a subtle bug with bindings involving
```
....? looks like the meat of this note is missing !

File: tests/src/core/CMakeLists.txt [hunk 8]
```diff
@@ -203,7 +204,7 @@ set(TESTS
  testqgsweakrelation.cpp
  testqgsziputils.cpp
  testqobjectuniqueptr.cpp
- testziplayer.cpp
+ testziplayer.cpp
```
Comment (nyalldawson) [inline tests/src/core/CMakeLists.txt:207]:
```suggestion
 testziplayer.cpp
```

File: tests/src/core/testqgsrelation.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+#include "qgsmaplayerstylemanager.h"
+#include "qgstest.h"
+#include "qgsvectorlayer.h"
+
+
+/**
+ * \ingroup UnitTests
+ * This is a unit test for the QgsRelation changing style
+ */
+class TestQgsRelation: public QObject
```
Comment (nyalldawson) [inline tests/src/core/testqgsrelation.cpp:31]: is this change related? Or an accidental commit?

File: tests/src/core/testqgsrelation.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
+/***************************************************************************
+     testqgssettings.cpp
```
Comment (nyalldawson) [inline tests/src/core/testqgsrelation.cpp:2]:
```suggestion
     testqgsrelation.cpp
```

File: src/gui/qgsattributeform.cpp [hunk 11]
```diff
@@ -2694,6 +2695,19 @@ void QgsAttributeForm::ContainerInformation::apply( QgsExpressionContext *expres
...
     isVisible = newVisibility;
   }
+
+  const bool newCollapsedState = collapsedExpression.evaluate( expressionContext ).toBool();
+
+  if ( collapsedExpression.isValid() && ! collapsedExpression.hasEvalError() && newCollapsedState != isCollapsed )
+  {
+
+    QgsCollapsibleGroupBoxBasic *collapsibleGroupBox { qobject_cast<QgsCollapsibleGroupBoxBasic *>( widget ) };
+    if ( collapsibleGroupBox )
```
Comment (nyalldawson) [inline src/gui/qgsattributeform.cpp:2705]:
```suggestion
    if ( QgsCollapsibleGroupBoxBasic *collapsibleGroupBox { qobject_cast<QgsCollapsibleGroupBoxBasic *>( widget ) } )
```

**Review by nyalldawson — APPROVED** [review]

File: tests/src/core/testqgsrelation.cpp (hunk 9 above)
Comment (elpaso) [reply tests/src/core/testqgsrelation.cpp:31]: It is related: this is a C++ implementation of the original test_qgsrelation.py without the reload changes that I had to introduce in the test.

Long story short: after the introduction of QgsOptionalExpression mCollapsedExpression in https://github.com/qgis/QGIS/pull/47671/files#diff-9a0282d186d8598b3a272358b7517b17b6350908ca79340accd62c5a86852b7dR217 the test_qgsrelation.py started failing, the change is apparently unrelated and cannot be reproduced when the test is run through the QGIS console so I've been running debugging session for hours and decided that it's a SIP issue somewhere that makes the test fail and it is not an issue in real use cases, to add further protection I have re-implemented the original test in C++ without the modifications that I had to make to the python test to make it pass.

File: tests/src/python/test_qgsrelation.py (hunk 7 above)
Comment (elpaso) [reply tests/src/python/test_qgsrelation.py:199]: Yeah, it was an explanation of https://github.com/qgis/QGIS/pull/47671/files#r825704937 I'll re-add it.

File: python/core/auto_generated/editform/qgsattributeeditorcontainer.sip.in (hunk 1 above)
Comment (elpaso) [reply python/core/auto_generated/editform/qgsattributeeditorcontainer.sip.in]: done
